modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala

```diff
@@ -0,0 +1,71 @@
+package org.tessellation.infrastructure.snapshot
+
+import cats.data.NonEmptyList
+import cats.effect.{Async, Ref}
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.option._
+import cats.syntax.order._
+import cats.syntax.traverse._
+
+import scala.util.control.NoStackTrace
+
+import org.tessellation.dag.snapshot.{GlobalSnapshot, SnapshotOrdinal, StateChannelSnapshotWrapper}
+import org.tessellation.domain.snapshot.GlobalSnapshotStorage
+import org.tessellation.ext.crypto._
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.address._
+import org.tessellation.security.hash.Hash
+
+object GlobalSnapshotStorage {
+
+  def make[F[_]: Async: KryoSerializer: Ref.Make](genesis: GlobalSnapshot): F[GlobalSnapshotStorage[F]] =
+    Ref[F].of[NonEmptyList[GlobalSnapshot]](NonEmptyList.of(genesis)).map(ref => make(ref))
+
+  def make[F[_]: Async: KryoSerializer: Ref.Make](
+    snapshotsRef: Ref[F, NonEmptyList[GlobalSnapshot]]
+  ): GlobalSnapshotStorage[F] = new GlobalSnapshotStorage[F] {
+
+    def add(globalSnapshot: GlobalSnapshot): F[Unit] =
+      for {
+        maybeError <- snapshotsRef.modify { snapshots =>
+          val lastSnapshot = snapshots.head
+          lastSnapshot.hash match {
+            case Left(error) => (snapshots, error.some)
+            case Right(lastSnapshotHash) =>
+              val expectedLink = (lastSnapshotHash, SnapshotOrdinal.next.next(lastSnapshot.ordinal))
+              val actualLink = (globalSnapshot.lastSnapshotHash, globalSnapshot.ordinal)
+              if (expectedLink === actualLink)
+                (globalSnapshot :: snapshots, none[Throwable])
+              else
+                (snapshots, InvalidGlobalSnapshotChain(expectedLink, actualLink).some)
+          }
+        }
+        _ <- maybeError.traverse(_.raiseError[F, Unit])
+      } yield ()
+
+    def get(ordinal: SnapshotOrdinal): F[Option[GlobalSnapshot]] =
+      snapshotsRef.get.map { snapshots =>
+        snapshots.find(_.ordinal === ordinal)
+      }
+
+    def getLast: F[GlobalSnapshot] = snapshotsRef.get.map(_.head)
+
+    def getStateChannelSnapshot(
+      address: Address,
+      ordinal: SnapshotOrdinal
+    ): F[Option[StateChannelSnapshotWrapper]] = snapshotsRef.get.map { snapshots =>
+      snapshots.find { snapshot =>
+        snapshot.ordinal <= ordinal && snapshot.stateChannelSnapshots.contains(address)
```

Comment:
I think that the method name may be misleading, I will rename it. The intention is to return the most recent state-channel snapshot that's not above passed ordinal. It may happen that in the GlobalSnapshot at ordinal 7 there will be no state-channel snapshot for a given address, so the most recent one will be earlier (if any).

Reply:
Ok, please change the name accordingly. I had the same thought as @TheMMaciek.